execution-engine/contracts/hdac-system/pop/src/contract_queue/request_queue.rs

```diff
@@ -0,0 +1,115 @@
+        RequestQueueEntry::<T> {
+            request_key,
+            amount,
+            timestamp,
+        }
+    }
+}
+
+pub trait RequestKey: Clone + Copy + PartialEq + FromBytes + ToBytes + CLTyped {}
+
+impl<T: RequestKey> RequestQueue<T> {
+    pub fn push(&mut self, request_key: T, amount: U512, timestamp: BlockTime) -> Result<()> {
+        if self.0.iter().any(|entry| entry.request_key == request_key) {
+            return Err(Error::MultipleRequests);
+        }
+        if let Some(entry) = self.0.last() {
+            if entry.timestamp > timestamp {
+                return Err(Error::TimeWentBackwards);
+            }
+        }
+        self.0.push(RequestQueueEntry {
+            request_key,
+            amount,
+            timestamp,
+        });
+        Ok(())
+    }
+    pub fn pop_due(&mut self, timestamp: BlockTime) -> Vec<RequestQueueEntry<T>> {
+        let (older_than, rest) = self
+            .0
+            .iter()
+            .partition(|entry| entry.timestamp <= timestamp);
+        self.0 = rest;
+        older_than
+    }
+}
+
+impl<T: RequestKey> FromBytes for RequestQueue<T> {
+    fn from_bytes(bytes: &[u8]) -> result::Result<(Self, &[u8]), bytesrepr::Error> {
+        let (len, mut bytes) = u64::from_bytes(bytes)?;
+        let mut queue = Vec::new();
+        for _ in 0..len {
+            let (entry, rest) = RequestQueueEntry::from_bytes(bytes)?;
+            bytes = rest;
+            queue.push(entry);
+        }
+        Ok((RequestQueue(queue), bytes))
+    }
+}
+
+impl<T: RequestKey> ToBytes for RequestQueue<T> {
+    fn to_bytes(&self) -> result::Result<Vec<u8>, bytesrepr::Error> {
+        let mut bytes = (self.0.len() as u64).to_bytes()?; // TODO: Allocate correct capacity.
+        for entry in &self.0 {
+            bytes.append(&mut entry.to_bytes()?);
+        }
+        Ok(bytes)
+    }
+}
+
+impl<T: RequestKey> CLTyped for RequestQueue<T> {
+    fn cl_type() -> CLType {
+        CLType::List(Box::new(RequestQueueEntry::<T>::cl_type()))
+    }
+}
+
+impl<T: RequestKey> FromBytes for RequestQueueEntry<T> {
+    fn from_bytes(bytes: &[u8]) -> result::Result<(Self, &[u8]), bytesrepr::Error> {
+        let (request_key, bytes) = T::from_bytes(bytes)?;
+        let (amount, bytes) = U512::from_bytes(bytes)?;
+        let (timestamp, bytes) = BlockTime::from_bytes(bytes)?;
+        let entry = RequestQueueEntry {
+            request_key,
+            amount,
+            timestamp,
+        };
+        Ok((entry, bytes))
+    }
+}
+
+impl<T: RequestKey> ToBytes for RequestQueueEntry<T> {
+    fn to_bytes(&self) -> result::Result<Vec<u8>, bytesrepr::Error> {
+        Ok((self.request_key.to_bytes()?.into_iter())
+            .chain(self.amount.to_bytes()?)
+            .chain(self.timestamp.to_bytes()?)
+            .collect())
+    }
+}
+
+impl<T: RequestKey> CLTyped for RequestQueueEntry<T> {
+    fn cl_type() -> CLType {
+        CLType::Any
```

Comment:
How about using `CLType::Tuple3`? 
If `Any` is used, it cannot be serialized and deserialized by `CLType`. 
And if it is changed to a format other than `Any`,
`ToBytes`, `FromBytes` and `CLTyped for RequestQueue<T>` must also be changed according to the `CLType` format.

Reply:
> If Any is used, it cannot be serialized and deserialized by CLType.
I don't understand this point. They are already serialized and deserialized through `ToBytes` and `FromBytes` as `CLType::Any`.

The `CLType`s other than `CLType::Any` provide `ToBytes` and `FromBytes` implementations in the sdk level and they also allow us to implement `CLType` for our own user-defined type as `CLType::Any`.

And `RequestQueueEntry` is a generic struct, not a tuple.
I recommend you to check [this](https://github.com/hdac-io/CasperLabs/blob/94b5be1f53eaf53af05c262e624b867a7deb06c7/execution-engine/types/src/cl_type.rs#L647) out.

Reply:
It doesn't matter if `RequestKey` is not used outside of execution-engine. If used externally, `CLType::Any` cannot be deserialized.
I think about it now, may not need external access because the required data can be created and thrown at the step function.
If only the key of the queue is used, it may not be necessary, but if possible, keep the serialized format for `CLType`.

Reply:
1. This is an internal data periodically flushed so it doesn't require deserialization outside of EE.
2. If we choose to use a tuple, then we cannot add associated functions or methods on that type and also we cannot bestow type safety.
3. The reason for implementing CLType in this context is to store the data into EE storage not to interact with a client. I think it is overkill to prohibit defining a struct as much as possible due to the possibility of interaction. I think it is enough to change them or create a new interface when the necessity occurs.

Reply:
1. 👌
2. I don't think so. You can use the generic struct and define the function for that struct. I think CLType is just the type for serialized.
3. The types stored in the store are saved by type, and shouldn't they be stored the same? 

There is no problem with the operation, so I followed.

Reply:
> 2. I don't think so. You can use the generic struct and define the function for that struct. I think CLType is just the type for serialized.

Nope. The `CLTyped` trait is just a trait describing what the type is.
The actual serialization and deserialization are performed by `ToBytes` and `FromBytes` trait.
So even if the `fn cl_type()` is implemented to return `CLType::Tuple3`, they are serialized and deserialized as implemented in `ToBytes` and `FromBytes`.

Reply:
Of course serialized, deserialized uses `ToBytes` and `FromBytes`.
Check the `stored value` section. Decide how to serialize each according to the `CLTyped`. It is said that you can define a type by combining `CLType` according to your struct.

Reply:
> It is said that you can define a type by combining CLType according to your struct.

It doesn't say we can define a struct that implements `CLTyped` with `CLType::Tuple3`.

I don't know what is the point of mentioning about `engine_shared::stored_value::StoredValue`. This enum is not implementing `CLTyped` nor a fitting example for this case. This is just an abstraction classifying the type of data stored in the ee storage.

We can define a struct consisting of `CLTyped` entries but we cannot store that struct directly into the ee storage.

To write the data into the ee storage, it requires the type of that data should be `CLTyped`. But if you force to use pre-defined CLType(which means not `CLType::Any`), this means we cannot store our own data type directly into the ee storage.

In other words, if we want to store our own struct type into the ee storage, we should implement `CLTyped` with `CLType::Any` for that struct.